### DISCUSSION TITLE
AMICO Maps histograms general question- Not an issue

### DISCUSSION THREAD

**Discussion by NeuroImagingWorld** [daducci/AMICO#173] [Q&A]
Hello Dr. Daducci,

  This is not an issue but general question in regard to histograms for the AMICO maps that are generated. There are spikes showing up at certain intervals for FICVF and OD histograms as attached. When we threshold the image for these values the data seem to be spread across the brain. 

Can you please advice if this is expected.

Thanks,
Prasanna

![amico_histograms](https://cloud.githubusercontent.com/assets/10892317/9421905/3769bfce-4833-11e5-9e01-1d62b3e54e20.gif)

**Comment by daducci** [comment]
Dear Prasanna,

I think this is an effect of the discretization in the dictionary. When you apply the l1 norm to promote sparsity in the solution, only some atoms are retained from all possible combinations that are considered.
Did you try to play with the regularization?

I had observed similar spikes also in the original NODDI fitting, especially in the OD parameter, where the values attributed to CSF is always the same.

Best,
Alessandro

**Comment by NeuroImagingWorld** [comment]
Hello,

   Attached are the comparison of NODDI and AMICO histograms. 

![screen shot 2015-08-28 at 4 31 17 pm](https://cloud.githubusercontent.com/assets/10892317/9559052/4508f20c-4da2-11e5-9f2f-1b506a95fec0.png)
![screen shot 2015-08-28 at 4 35 16 pm](https://cloud.githubusercontent.com/assets/10892317/9559101/15ff2bba-4da3-11e5-9a9f-6a5119e1338b.png)

Here are the voxels associated with corresponding spikes for ICVF from AMICO results

![screen shot 2015-08-28 at 4 36 10 pm](https://cloud.githubusercontent.com/assets/10892317/9559107/245d2ad6-4da3-11e5-903b-ebc1cc3db30c.png)
![screen shot 2015-08-28 at 4 36 29 pm](https://cloud.githubusercontent.com/assets/10892317/9559108/245d2194-4da3-11e5-9e54-d50843779fa4.png)
![screen shot 2015-08-28 at 4 36 43 pm](https://cloud.githubusercontent.com/assets/10892317/9559105/24562c90-4da3-11e5-8e44-29a2344e10c2.png)
![screen shot 2015-08-28 at 4 36 52 pm](https://cloud.githubusercontent.com/assets/10892317/9559106/2459db42-4da3-11e5-8728-809c737c5d31.png)
![screen shot 2015-08-28 at 4 37 06 pm](https://cloud.githubusercontent.com/assets/10892317/9559109/246307ee-4da3-11e5-82cf-ef1334249f16.png)

We have tried adjusting the regularization parameters and following are the results.

![screen shot 2015-08-28 at 4 43 06 pm](https://cloud.githubusercontent.com/assets/10892317/9559166/e4d917b6-4da3-11e5-83f6-c956bdee975f.png)
![screen shot 2015-08-28 at 4 48 20 pm](https://cloud.githubusercontent.com/assets/10892317/9559212/b287a510-4da4-11e5-9170-619b55a9501f.png)

Can you please share your thoughts. 

Thanks,
Prasanna

**Comment by daducci** [comment]
Dear Prasanna,

sorry for my delay, I am at the moment on kind of "paternity leave" and have had not much time to work. Sorry for that; I'm trying to catch up with work as fast as I can!

What you highlight seems to me related to the discretization AMICO uses  in the dictionary. I think the spikes are generated by this discretization that somehow limits the possible output values for the parameters. These values seem in fact widely spread in the brain, and not concentrated in any specific location. Surely there are differences between the original model and the aMICO approximation (as I previously described), but this is another topic.

Do you think this could badly influence your analyses?

On a side note, we have recently developed an alternative way to fit models within AMICO that can enforce, for free, spatial regularization of the estimations in each voxel; this might help also with the issue you describe here. We have not tested this method yet, but I will speak with @ejcanalesr (the actual person who developed this new solver) and I will discuss with him about its applicability to mitigate the discretization issue you have described.

I will keep you posted!
Best,
Ale

**Comment by NeuroImagingWorld** [comment]
Hello Dr. Daducci,

  Thank you for looking into this. We are comparing the response from both NODDI and AMICO to determine if current NODDI processing can be replaced with AMICO  as this is the faster version of NODDI or if there are areas that we need to be cautious about in analyzing the results in tumor regions in addition to normal brain processing. 

Will keep you posted once we do quantitative analysis in tumor regions. Appreciate your feedback.

Thanks,
Prasanna

**Comment by daducci** [comment]
Dear Prasanna,

did you finally got any useful information from the comparison between AMICO and NODDI implementations?

**Comment by NeuroImagingWorld** [comment]
Hello Dr. Daducci,

```
Thank you for following up on this item. Based on our preliminary analysis we did notice that FICVF values are similar between NODDI and AMICO in both tumor and white mater regions. However, in some cases, ODI and FISO values from AMICO seem to be higher in tumor region compared to NODDI results thus effecting the correlations with histopathology.
```

Can you please suggest if there are any parameters we could adjust to make these two results match. 

Thanks,
Prasanna

**Comment by steelec** [comment]
To follow up on this, I have run into the same issue with discretization in the ICVF and ODI (running the python version of AMICO). Has there been any progress on optimal paramteres or the method by @ejcanalesr?
Chris
Screenshot of AMICO calculated OD versus that calculated by the NODDI matlab toolbox. The discretizations appear at values very close to .19,.29,.39...
![screenshot from 2016-04-06 20 23 42](https://cloud.githubusercontent.com/assets/2734273/14360361/fe956950-fcc3-11e5-8e52-012a451fa7eb.png)

same comparison, with mean gaussian smoothing of the AMICO OD ouptut, kernel size from left to right = 0.75, 1, 2mm (data resolution = 2x2x2)

![screenshot from 2016-04-07 13 40 22](https://cloud.githubusercontent.com/assets/2734273/14360997/db57131e-fcc6-11e5-8475-1c6277c708b0.png)

**Comment by NeuroImagingWorld** [comment]
Hello Dr. Dadduci,

  Can you please point us to the method proposed by @ejcanalesr, we can plan to include it in further analysis.

Thanks,
Prasanna

**Comment by ejcanalesr** [comment]
Hello Dr. Prasanna,
The method is not ready yet, however, I will have time to work on it in
july.
I will try to have a final solution as soon as possible.
Thanks,
Regards,

_Erick J. Canales-Rodríguez_
FIDMAG Germanes Hospitalàries
CIBERSAM
Telf.: 93 652 99 99 ext. 1493

_www.fidmag.org http://www.fidmag.org/_

Ley de protección de datos:
La información contenida en este e-mail es confidencial y privada, puede
contener información relativa a datos especialmente protegidos por la Ley
Orgánica 5/1999 de Protección de los Datos de Carácter Personal. Si usted
no es la persona a la que iba dirigido este mensaje, por favor, contacte
inmediatamente con el remitente. Cualquier copia o uso no autorizado de
este e-mail queda prohibido.

2016-06-12 18:45 GMT+02:00 goals2008 notifications@github.com:

**Comment by daducci** [comment]
Dear Prasanna,

we have started looking into this issue and we have a question for you. We noticed that in your plots the OD index is about 0.7 whereas in the maps released with the tutorial data OD is about 1. Which version of the code and matlab are you using?

Thanks,
Ale

**Comment by NeuroImagingWorld** [comment]
Hello,

   Thank you for looking into this issue. We are using following Matlab
2013b..and default version of AMICO. Also ODI is about 1 in all the maps. I
think 0.7 maps that were showing up is with L2 norm comparison, screen seem
to have been cut off.

Please let me know if you need more information.

Thanks,
Prasanna

**Comment by ejcanalesr** [comment]
Dear Prasanna,

We have been working on the issue you reported. Fortunately, we have found
a potential solution to obtain very similar reconstructions but without the
spikes you reported in the histograms. It involves only a small correction
in the debiasing step of the estimation, so the numerical algorithm is
totally compatible with the one reported in our article.

As we have tested this preliminar version only in our data (as well as in
synthetic data), we would like to test the method also in different
datasets. This step is important to determine whether the new approach is
robust and stable.

Please, could you send us the data you have used to create the figures
reported in the web? Or alternatively, if you prefer, we can send you this
preliminar version of the code so that you could test the method in your
computer and generate similar figures. This may be possible?

We plan to update the toolbox as soon as possible, after verify the
robustness and stability of the new version.

Thank you very much for your valuable contribution!

Regards,

P.S. Please, find attached here a preliminary result from our evaluation
with the new method. These maps were computed from the example dataset
distributed in the original NODDI toolbox, but we have carried out the
estimation in the whole brain.

Notice that after recomputing the OD maps with the original NODDI, in CSF
we obatin OD = 0.7, instead of the value OD = 1 resulting from the previous
version of the sofware. In our method we have OD=1 in CSF...
_Erick J. Canales-Rodríguez_
FIDMAG Germanes Hospitalàries
CIBERSAM
Telf.: 93 652 99 99 ext. 1493

_www.fidmag.com http://www.fidmag.com/_

**Comment by ejcanalesr** [comment]
![od_whole_brain_new_hists](https://cloud.githubusercontent.com/assets/9079725/17105887/5ca4f176-5289-11e6-8803-066dcb1ffbb0.jpg)

**Comment by NeuroImagingWorld** [comment]
Sure, I will be happy to test this out, please do share the updated code. I dont have access to the original data right now, but could send you another dataset soon.

Thanks,
Prasanna

**Comment by steelec** [comment]
This looks great - I would be happy to test as well on HCP data and a 3-shell 2mm dataset that we have collected in-house. I have been using the python version of AMICO, but would also be open to testing the matlab version.

Chirs

**Comment by daducci** [comment]
Thanks @goals2008 and @steelec for your help!
I think we could proceed in two steps (if you agree):
1. We test internally the new code and fine tune it, if possible using your data to compare with your previous results (top of this post).
2. Once these tests are done, we release the new code and you test it on the rest of your datasets to evaluate its effectiveness.

Sounds good?

**Comment by NeuroImagingWorld** [comment]
Yes, that sounds good.

Thanks
Prasanna

**Comment by steelec** [comment]
Perfect.

On Fri, Jul 29, 2016, 13:56 goals2008 notifications@github.com wrote:

> Yes, that sounds good.
> 
> Thanks
> Prasanna
> 
> > On Jul 29, 2016, at 4:39 AM, Alessandro Daducci <
> > notifications@github.com> wrote:
> > 
> > Thanks @goals2008 and @steelec for your help!
> > I think we could proceed in two steps (if you agree):
> > 
> > We test internally the new code and fine tune it, if possible using your
> > data to compare with your previous results (top of this post).
> > 
> > Once these tests are done, we release the new code and you test it on
> > the rest of your datasets to evaluate its effectiveness.
> > 
> > Sounds good?
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/daducci/AMICO/issues/8#issuecomment-236162142, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/ACm4wbcy75Lcz0nfTRg4yC6sqebH_wCtks5qaepbgaJpZM4FwQyD
> .
> 
> ## 

Chris

**Comment by NeuroImagingWorld** [comment]
Here is the test data having similar results. Please let us know if you see
any issues.

https://www.dropbox.com/s/xlxnu1hoaxciuwa/TestData_MASI.zip?dl=0

Thanks,
Prasanna

**Comment by NeuroImagingWorld** [comment]
Hello Dr. Daducci,

   Can you please let us know if there is any additional information you are looking for or if there is any testing we could do for this fix.

Thanks,
Prasanna

**Comment by ejcanalesr** [comment]
Hi, we have finished the internal evaluation step (first step proposed by @daducci in the previous post) by using different datasets (thanks @goals2008 and @steelec!).
Please find attached here some results. As can be seen, the histograms look smoother without large spikes. So, we think we are ready for the next step.  @goals2008 and @steelec, are you able to test the new code on the rest of your datasets? Thanks for your help!

Result from the "example dataset" included as part of the original NODDI toolbox
(the whole-brain reconstruction time of the new method was 16 minutes).

![icvf](https://cloud.githubusercontent.com/assets/9079725/18274610/899f0afa-7443-11e6-90bd-c2186015d93a.jpg)
![iso](https://cloud.githubusercontent.com/assets/9079725/18274609/899c6886-7443-11e6-8e70-c0cde35e8edf.jpg)
![od](https://cloud.githubusercontent.com/assets/9079725/18274611/89aeed9e-7443-11e6-8221-deedf5465811.jpg)

**Comment by steelec** [comment]
Great to hear!
I would be happy to test. Have the changes been committed to the master or another branch?
Cheers,
Chris

**Comment by NeuroImagingWorld** [comment]
Hello Canales,

```
Thanks for the update, will be happy to test it out. Can we get the updated code.
```

Thanks
Prasanna

**Comment by ejcanalesr** [comment]
Dear @goals2008 and @steelec, please find attached here the new code.

Short explanation (MATLAB toolbox):
1) Rename the "old" AMICO_NODDI function to "AMICO_NODDI_old"  (a new version of this function is included in the attached folder)
2) Unzip and add all new functions to the matlab path
3) The estimation is carried out as in the tutorial: https://github.com/daducci/AMICO/tree/master/matlab/doc/demos/NODDI
HOWEVER, PLEASE NOTE: to fit the AMICO-NODDI model using the new implementation run the command line: 
AMICO_NewFit( ) (instead of the old command: AMICO_Fit()).

The readme.txt file contains more details...

[new_AMICO_NODDI_scripts_07_09_2016.zip](https://github.com/daducci/AMICO/files/459322/new_AMICO_NODDI_scripts_07_09_2016.zip)

**Comment by NeuroImagingWorld** [comment]
Thank you for providing the updated code and detailed instructions. Will
test this out and keep you posted.

Thanks,
Prasanna

**Comment by steelec** [comment]
I can do the same ASAP if you also have a python version, otherwise I can probably get to testing the matlab version sometime late next week.

Thanks very much!

**Comment by daducci** [comment]
Hi @goals2008 and @steelec ,

the code is just an "internal test" for the moment, so it's been prototyped in MATLAB to have a quick version asap (we are more familiar with MATLAB than python). Then, if you can confirm the improved results we have observed, we will proceed to porting it to python as well.

Thanks a lot for you help!

**Comment by NeuroImagingWorld** [comment]
Hi Erick,

   Sorry for the delay. I had to re-setup my AMICO/spams Matlab environment
as I was using python version as well.

I got it running for both original and new model and results are looking
good as attached. Thank you for making this fix.

 Thanks,
Prasanna

**Comment by steelec** [comment]
I don't think it was attached...?
Sorry for my slowness on testing as well, probably will be able to do some more testing in about 2 weeks.
Chris

**Comment by NeuroImagingWorld** [comment]
I have attached 2 screenshots having ICVF and ODI histogram and qualitative
image
- AMICO_Orig
- AMICO_New

Reattaching them.. can you please confirm if you still see issue accessing
them.

Thanks,
Prasanna

**Comment by steelec** [comment]
I don't see it here. weird.

**Comment by daducci** [comment]
I don't see any attachment either.

**Comment by NeuroImagingWorld** [comment]
Attaching it again here.. may be its an issue when sending from email. 
![amico_new](https://cloud.githubusercontent.com/assets/10892317/19418328/52fd9134-9387-11e6-9a54-e06ea77eb860.png)
![amico_orig](https://cloud.githubusercontent.com/assets/10892317/19418329/557d084a-9387-11e6-8d05-e0c483cc11cc.png)

**Comment by ejcanalesr** [comment]
Hi goals2008,
Please, could you provide more details about these figures to identify the outputs from the original and new reconstructions?
Thank you very much!!!

**Comment by NeuroImagingWorld** [comment]
Yes.
--> Top row is from AMICO_New for ODI and ICVF maps (No Spikes)
-->  Bottom row is from AMICO_orig for ICVF and ODI (with spikes)

Thanks,
Prasanna

**Comment by NeuroImagingWorld** [comment]
Hi Erick,

  Just checking if the fix you have made earlier is available in python
version or in the official matlab version.

Appreciate your efforts.

Thanks,
Prasanna

**Comment by ejcanalesr** [comment]
Dear Prasanna,

The fix is not available in the official release yet.

However, we are working in a new different implementation that includes a Rician noise model, and spatial regularization for noise datasets, which may replace that fix.

We are very close to have the final version but still we need some time to test it in different datasets.

Thank you very much for your interest!

Regards,

Erick



________________________________

**Comment by NeuroImagingWorld** [comment]
Thank you Erick! Its great to know that it will be available soon!

**Comment by jling-NM** [comment]
This issue is about to reach its anniversary. 
Has any progress been made on bringing this new fix to the python code base? We would really like to use this work for a study but not until the spiking issue has been resolved.
Alternatively, do you suggest we go back to the old matlab code if we want to use AMICO?

thanks and please let us know how we can help.

**Comment by steelec** [comment]
Bump -- any progress on this?

**Comment by captainnova** [comment]
Hi, I'm also curious whether there has been any progress on this, especially in the python version. I greatly appreciate the speed and stability of the AMICO implementation of NODDI, but it seems as if it is sometimes too stable and prefers certain values. We would like to make comparisons within a set of scans for a study, but as you can see we have a problem with the ICVF (= NDI in the plot) having a predilection for certain values for many regions in the JHU "Eve" WM atlas:

![jhumediannoddi_ndi](https://user-images.githubusercontent.com/15852777/45908330-afc41080-bdc1-11e8-9f13-2918db0bef11.png)

The plot shows the median ICVF of the voxels inside each atlas region vs. the number of voxels in that scans's region in subject space for 128 scans. I like to make these plots to check the quality of both model fitting and atlas registration, and usually expect to see blobbish clusters for each region. The orientation dispersion and ISOVF from AMICO NODDI are not showing any drastic quantization in their plots. 

Is there any way I can help? If it is delayed by the Rician noise model, I hope you will consider making a release without it, since I am already debiasing my input images.

Thanks,

Rob Reid

**Comment by jling-NM** [comment]
My personal impression is that this issue cannot or will not be fixed.

If you are looking for faster modelling of NODDI parameters and know how to setup GPU devices you may explore the [Microstructure Diffusion Toolbox](https://mdt-toolbox.readthedocs.io/en/latest_release/index.html)

If you scroll half way down the [FAQ](https://mdt-toolbox.readthedocs.io/en/latest_release/faq.html) page you'll see an example of NODDI modelling and their publication.
I ran a test on a single subject with a basic NVIDIA card and the total run time was ~5 minutes.

An added benefit is the software allows testing many models and easily adding your own so it is an interesting way to explore your data and learn concepts if you are so inclined.

**Comment by captainnova** [comment]
> If you are looking for faster modelling of NODDI parameters and know how to setup GPU devices you may explore the [Microstructure Diffusion Toolbox](https://mdt-toolbox.readthedocs.io/en/latest_release/index.html)

Thanks! We are doing our diffusion processing with GPUs anyway because of eddy_cuda, but do have a bit of a road bump because the Microstructure Diffusion Toolbox requires python 3. We are using CentOS 7, and I know there are a few ways to install python 3 even on RHEL-based systems, but I will have to work with our system administrator a bit to figure something out.

**Comment by daducci** [comment]
Dear @captainnova ,

indeed we have to apologize for the long silence!

> My personal impression is that this issue cannot or will not be fixed.

The good news is that the problem can be solved, and in fact it was! We have had the code ready for long time now, but have no one to port it and integrate to AMICO. We will do our best to port it asap.

As a side note, we are working also on porting AMICO to GPU, for an event more dramatic acceleration! We'll keep you posted of course.

**Comment by captainnova** [comment]
> The good news is that the problem can be solved, and in fact it was! We have had the code ready for long time now, but have no one to port it and integrate to AMICO. We will do our best to port it asap.
> 
> As a side note, we are working also on porting AMICO to GPU, for an event more dramatic acceleration! We'll keep you posted of course.

I'm glad to hear it!  

Thanks, Rob

**Comment by stremblay18** [comment]
Dear @daducci , 

I am having the same issue with the discretization of ICVF values (I'm working on the HCP dataset). 

Is this implemented into the Python version of AMICO? 

Or is there a workaround I could use in the meantime (increasing the  number of evenly spaced samples in np.linspace(0.1, 0.99, 12) from 12 to 50 maybe?) as mentioned in issue #97 https://github.com/daducci/AMICO/issues/97 for a similar issue.

<img width="492" alt="Screen Shot 2021-06-23 at 5 03 08 PM" src="https://user-images.githubusercontent.com/52012786/123167715-e896e580-d444-11eb-8c60-02b664c0d779.png">

I would greatly appreciate any help!

Thank you,

Stefanie

**Comment by daducci** [comment]
Dear Stefanie,
in the end, this fix has never been implemented as it turned out that this "discretization problem" is not really an issue. In fact, it is true that AMICO tends to favour some configutations w.r.t. others, but the values are very close and differences are really mild. In any case, I will try to touch base with @ejcanalesr to see if he can implement this so, perhaps, we could finally test whether indeed these discretization spikes do have or do not have any impact on the results.
